cache/lib/cache_web/controllers/module_cache_controller.ex

```diff
@@ -0,0 +1,331 @@
+defmodule CacheWeb.ModuleCacheController do
+  use CacheWeb, :controller
+  use OpenApiSpex.ControllerSpecs
+
+  alias Cache.BodyReader
+  alias Cache.CacheArtifacts
+  alias Cache.Disk
+  alias Cache.S3
+  alias Cache.S3Transfers
+  alias CacheWeb.API.Schemas.Error
+
+  require Logger
+
+  plug OpenApiSpex.Plug.CastAndValidate, json_render_error_v2: true
+
+  tags(["ModuleCache"])
+
+  operation(:download,
+    summary: "Download a module cache artifact",
+    operation_id: "downloadModuleCacheArtifact",
+    parameters: [
+      project_id: [
+        in: :query,
+        type: :string,
+        required: true,
+        description: "Project ID (account_handle/project_handle)"
+      ],
+      hash: [
+        in: :query,
+        type: :string,
+        required: true,
+        description: "Artifact hash"
+      ],
+      name: [
+        in: :query,
+        type: :string,
+        required: true,
+        description: "Artifact name"
+      ],
+      cache_category: [
+        in: :query,
+        type: :string,
+        required: false,
+        description: "Cache category (builds, selective_tests)"
+      ]
+    ],
+    responses: %{
+      ok: {"Artifact content", "application/octet-stream", nil},
+      not_found: {"Artifact not found", "application/json", Error},
+      unauthorized: {"Unauthorized", "application/json", Error},
+      forbidden: {"Forbidden", "application/json", Error},
+      bad_request: {"Bad request", "application/json", Error}
+    }
+  )
+
+  def download(conn, params) do
+    with {:ok, {account_handle, project_handle}} <- parse_project_id(params),
```

Comment:
`parse_project_id` is a good candidate to be a function plug in this controller that turns `account_handle` and `project_handle` into assigns that can be read with pattern-matching in the function definition of every action.

Reply:
in the `server`, we usually have an `account_handle` and `project_handle` as separate parameters and parse the full handle to the account and project handles in the client. I wonder if we should align with that approach here.